File: Java/MongoTodo/src/main/java/com/mug/todo/TodoItemDAO.java
package com.mug.todo;

import java.util.*;

import org.bson.types.ObjectId;
import org.springframework.stereotype.Repository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.mongodb.core.query.Update;

@Repository
public class TodoItemDAO{
	@Autowired
    private MongoTemplate mongoTempl;
	
	public void insert(TodoItem item){
		item.setAuthor("Mrs. Java");
		item.setDone(false);
		item.setDate(new Date());
		mongoTempl.insert(item,"todoItems");
	}
	
	public List<TodoItem> getAll()
	{
		return mongoTempl.find(new Query().with(
				new Sort(Sort.Direction.DESC, "date")),
				TodoItem.class, "todoItems");
	}
	public List<TodoItem> getByStatus(boolean status)
	{
		return mongoTempl.find(
				new Query(Criteria.where("isDone").is(status)).with(new Sort(Sort.Direction.DESC, "date")), TodoItem.class, "todoItems");
	}
	public List<TodoItem> getByTag(String tag)
	{
		return mongoTempl.find(new Query(Criteria.where("tags").is(tag)).with(new Sort(Sort.Direction.DESC, "date")), TodoItem.class, "todoItems");
	}
	public void changeStatusDone(String elementId)
	{
		mongoTempl.updateFirst(
				new Query(Criteria.where("_id").is(new ObjectId(elementId))), new Update().set("isDone", true), TodoItem.class, "todoItems");
	}
	public void delete(String elementId)
	{
		mongoTempl.remove(new Query(Criteria.where("_id").is(new ObjectId(elementId))), TodoItem.class, "todoItems");
	}
}

File: Python/todoItemDAO.py
import datetime
from bson.objectid import ObjectId

class TodoItemDAO:
    def __init__(self, database):
        self.db = database
        self.collection = database.todoItems

    def insert(self, title, tags):
        todoItem = {"title":title,
                    "author":"Sir Python",
                    "isDone": False,
                    "tags":tags,
                    "date": datetime.datetime.utcnow()}
        return self.collection.insert(todoItem)

    def get(self):
        return list(self.collection.find().sort("date",-1))
        
    def get_by_status(self, isDone):    
        return list(self.collection.find({"isDone":isDone}).sort("date",-1))
    
    def get_by_tag(self, tag):
        return list(self.collection.find({"tags":tag}).sort("date",-1))
    
    def change_status_to_done(self, item_id):
        self.collection.update({"_id": ObjectId(str(item_id))}, {"$set":{"isDone":True}})
    
    def delete(self, item_id):
        self.collection.remove({"_id": ObjectId(str(item_id))})

        
        
            
        
        
    
